tsdb/exemplar.go

```diff
@@ -0,0 +1,209 @@
+
+package tsdb
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type CircularExemplarStorage struct {
+	lock        sync.RWMutex
+	index       map[string]int
+	exemplars   []*circularBufferEntry
+	nextIndex   int
+	len         int
+	secondaries []storage.ExemplarAppender
+}
+
+type circularBufferEntry struct {
+	se   storageExemplar
+	prev int // index of previous exemplar in circular for the same series, use -1 as a default for new entries
+}
+
+func (ce circularBufferEntry) Exemplar() exemplar.Exemplar {
+	return ce.se.exemplar
+}
+
+type storageExemplar struct {
+	exemplar        exemplar.Exemplar
+	seriesLabels    labels.Labels
+	scrapeTimestamp int64
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int, secondaries ...storage.ExemplarAppender) *CircularExemplarStorage {
+	return &CircularExemplarStorage{
+		exemplars:   make([]*circularBufferEntry, len),
+		index:       make(map[string]int),
+		len:         len,
+		secondaries: secondaries,
+	}
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+// TODO: separate wrapper struct for queries?
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+func (ce *CircularExemplarStorage) SetSecondaries(secondaries ...storage.ExemplarAppender) {
+	ce.secondaries = secondaries
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(start, end int64, l labels.Labels) ([]exemplar.Exemplar, error) {
+	var (
+		ret []exemplar.Exemplar
+		e   exemplar.Exemplar
+		idx int
+		ok  bool
+		buf []byte
+	)
+
+	ce.lock.RLock()
+	defer ce.lock.RUnlock()
+
+	if idx, ok = ce.index[l.String()]; !ok {
+		return nil, nil
+	}
+	lastTs := ce.exemplars[idx].se.scrapeTimestamp
+
+	for {
+		// We need the labels check here in case what was the previous exemplar for the series
+		// when the exemplar from the last loop iteration was written has since been overwritten
+		// with an exemplar from another series.
+		// todo (callum) confirm if this check is still needed now that adding an exemplar should
+		// update the index and previous pointer for the series whose exemplar was overwritten.
+		if idx == -1 || string(ce.exemplars[idx].se.seriesLabels.Bytes(buf)) != string(l.Bytes(buf)) {
+			break
+		}
+
+		e = ce.exemplars[idx].Exemplar()
+		// todo (callum) This line is needed to avoid an infinite loop, consider redesign of buffer entry struct.
+		if ce.exemplars[idx].se.scrapeTimestamp > lastTs {
+			break
+		}
+
+		lastTs = ce.exemplars[idx].se.scrapeTimestamp
+		// Prepend since this exemplar came before the last one we appeneded chronologically.
```

Comment:
Prepend is expensive as you have to move things around on each operation, so n^2. I'd suggest reversing as a separate final step which would be n.